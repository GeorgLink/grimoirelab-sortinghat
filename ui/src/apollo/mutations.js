import gql from "graphql-tag";

const LOCK_INDIVIDUAL = gql`
  mutation LockIndividual($uuid: String!) {
    lock(uuid: $uuid) {
      uuid
      individual {
        isLocked
      }
    }
  }
`;

const UNLOCK_INDIVIDUAL = gql`
  mutation UnlockIndividual($uuid: String!) {
    unlock(uuid: $uuid) {
      uuid
      individual {
        isLocked
      }
    }
  }
`;

const DELETE_IDENTITY = gql`
  mutation DeleteIdentity($uuid: String!) {
    deleteIdentity(uuid: $uuid) {
      uuid
    }
  }
`;

const MERGE = gql`
  mutation Merge($fromUuids: [String!], $toUuid: String!) {
    merge(fromUuids: $fromUuids, toUuid: $toUuid) {
      uuid
      individual {
        isLocked
        identities {
          name
          source
          email
          uuid
          username
        }
        profile {
          name
          id
        }
        enrollments {
          organization {
            name
          }
        }
      }
    }
  }
`;

const UNMERGE = gql`
  mutation unmerge($uuids: [String!]) {
    unmergeIdentities(uuids: $uuids) {
      uuids
      individuals {
        profile {
          name
          id
          isBot
        }
        identities {
          name
          source
          email
          uuid
          username
        }
        enrollments {
          start
          end
          organization {
            name
          }
        }
      }
    }
  }
`;

const MOVE_IDENTITY = gql`
  mutation moveIdentity($fromUuid: String!, $toUuid: String!) {
    moveIdentity(fromUuid: $fromUuid, toUuid: $toUuid) {
      uuid
      individual {
        isLocked
        identities {
          name
          source
          email
          uuid
          username
        }
        profile {
          name
          id
        }
        enrollments {
          organization {
            name
          }
        }
      }
    }
  }
`;

const lockIndividual = (apollo, uuid) => {
  let response = apollo.mutate({
    mutation: LOCK_INDIVIDUAL,
    variables: {
      uuid: uuid
    }
  });
  return response;
};

const unlockIndividual = (apollo, uuid) => {
  let response = apollo.mutate({
    mutation: UNLOCK_INDIVIDUAL,
    variables: {
      uuid: uuid
    }
  });
  return response;
};

const deleteIdentity = (apollo, uuid) => {
  let response = apollo.mutate({
    mutation: DELETE_IDENTITY,
    variables: {
      uuid: uuid
    }
  });
  return response;
};

const merge = (apollo, fromUuids, toUuid) => {
  let response = apollo.mutate({
    mutation: MERGE,
    variables: {
      fromUuids: fromUuids,
      toUuid: toUuid
    }
  });
  return response;
};

const unmerge = (apollo, uuids) => {
  let response = apollo.mutate({
    mutation: UNMERGE,
    variables: {
      uuids: uuids
    }
  });
  return response;
};

const moveIdentity = (apollo, fromUuid, toUuid) => {
  let response = apollo.mutate({
    mutation: MOVE_IDENTITY,
    variables: {
      fromUuid: fromUuid,
      toUuid: toUuid
    }
  });
  return response;
};

export {
  lockIndividual,
  unlockIndividual,
  deleteIdentity,
  merge,
  unmerge,
  moveIdentity
};
