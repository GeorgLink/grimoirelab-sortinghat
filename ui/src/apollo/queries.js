import gql from "graphql-tag";

const GET_INDIVIDUAL_BYUUID = gql`
  query GetIndividual($uuid: String!) {
    individuals(filters: { uuid: $uuid }) {
      entities {
        mk
        identities {
          name
        }
        profile {
          id
          name
        }
      }
    }
  }
`;

const GET_INDIVIDUALS = gql`
  query GetIndividuals($page: Int!, $pageSize: Int!) {
    individuals(page: $page, pageSize: $pageSize) {
      entities {
        mk
        isLocked
        identities {
          name
          source
          uuid
        }
        profile {
          id
          name
        }
      }
      pageInfo {
        page
        pageSize
        numPages
        hasNext
        hasPrev
        startIndex
        endIndex
        totalResults
      }
    }
  }
`;

const GET_PROFILE_BYUUID = gql`
  query GetProfileByUuid($uuid: String!) {
    individuals(filters: { uuid: $uuid }) {
      entities {
        isLocked
        profile {
          name
        }
        identities {
          name
          source
          email
          uuid
          username
        }
        enrollments {
          start
          end
          organization {
            name
          }
        }
      }
    }
  }
`;

const getIndividualByUuid = (apollo, uuid) => {
  let response = apollo.query({
    query: GET_INDIVIDUAL_BYUUID,
    variables: {
      uuid: uuid
    }
  });
  return response;
};

const getIndividuals = {
  currentPage: 1,
  numPages: 1,
  pageSize: 10,
  query: async function(apollo, pageSize) {
    let self = this;
    if (pageSize) {
      self.pageSize = pageSize;
    }
    let response = await apollo.query({
      query: GET_INDIVIDUALS,
      variables: {
        page: self.currentPage,
        pageSize: pageSize
      }
    });
    self.numPages = response.data.individuals.pageInfo.numPages;
    if (self.currentPage > self.numPages) {
      return undefined;
    }
    self.currentPage++;
    return response;
  }
};

const getProfileByUuid = (apollo, uuid) => {
  let response = apollo.query({
    query: GET_PROFILE_BYUUID,
    variables: {
      uuid: uuid
    }
  });
  return response;
};

export { getIndividuals, getIndividualByUuid, getProfileByUuid };
