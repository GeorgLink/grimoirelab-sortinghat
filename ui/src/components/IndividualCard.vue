<template>
  <v-card
    class="mx-auto"
    :class="{
      dropzone: isDragging,
      selected: isSelected,
      highlighted: isHighlighted
    }"
    raised
    v-on="$listeners"
    @drop.native.prevent.stop="onDrop($event)"
    @dragover.prevent.stop="isDragging = true"
    @dragenter.prevent.stop="isDragging = true"
    @dragleave="isDragging = false"
    @click="selectIndividual"
  >
    <v-list-item class="grow" three-line>
      <v-list-item-avatar color="grey lighten-2" size="30px">
        <span class="grey--text text--darken-3">{{ getNameInitials }}</span>
      </v-list-item-avatar>

      <v-list-item-content>
        <v-list-item-title class="font-weight-medium">
          {{ name || identities[0].identities[0].email }}
        </v-list-item-title>
        <v-list-item-subtitle v-if="enrollments && enrollments.length > 0">
          {{ enrollments[0].organization.name }}
        </v-list-item-subtitle>
        <v-list-item-subtitle>
          <v-icon
            v-for="source in sources"
            :key="source.name"
            v-text="selectSourceIcon(source)"
            small
            left
          />
        </v-list-item-subtitle>
      </v-list-item-content>

      <v-list-item-icon>
        <v-menu offset-y offset-x :close-on-content-click="false">
          <template v-slot:activator="{ on }">
            <v-btn icon v-on="on" @mousedown.stop>
              <v-icon small>
                mdi-magnify-plus-outline
              </v-icon>
            </v-btn>
          </template>
          <expanded-individual
            compact
            :enrollments="enrollments"
            :identities="identities"
            :uuid="uuid"
          />
        </v-menu>
        <v-btn text icon @click.stop="$emit('remove')" @mousedown.stop>
          <v-icon small>
            mdi-close
          </v-icon>
        </v-btn>
      </v-list-item-icon>
    </v-list-item>
    <slot />
  </v-card>
</template>

<script>
import ExpandedIndividual from "./ExpandedIndividual";

export default {
  name: "individualcard",
  components: {
    ExpandedIndividual
  },
  props: {
    name: {
      type: String,
      required: false,
      default: null
    },
    sources: {
      type: Array,
      required: false,
      default: () => []
    },
    isSelected: {
      type: Boolean,
      required: false,
      default: false
    },
    uuid: {
      type: String,
      required: true
    },
    identities: {
      type: Array,
      required: false
    },
    enrollments: {
      type: Array,
      required: false
    },
    isHighlighted: {
      type: Boolean,
      required: false,
      default: false
    }
  },
  data() {
    return {
      isDragging: false
    };
  },
  computed: {
    getNameInitials: function() {
      const name = this.name || this.identities[0].identities[0].email || "";
      const names = name.split(" ");
      let initials = names[0].substring(0, 1).toUpperCase();

      if (names.length > 1) {
        initials += names[names.length - 1].substring(0, 1).toUpperCase();
      }

      return initials;
    }
  },
  methods: {
    selectSourceIcon(source) {
      const datasource = source.toLowerCase();

      if (datasource === "others") {
        return "mdi-account-multiple";
      } else {
        return `mdi-${datasource}`;
      }
    },
    onDrop(event) {
      const type = event.dataTransfer.getData("type");
      if (type === "move") {
        this.moveIndividual(event);
      } else if (type === "enrollFromOrganization") {
        this.enrollIndividual(event);
      } else {
        this.mergeIndividuals(event);
      }
      this.isDragging = false;
    },
    selectIndividual() {
      this.$emit("select");
    },
    moveIndividual(event) {
      const uuid = event.dataTransfer.getData("uuid");
      this.$emit("move", { fromUuid: uuid, toUuid: this.uuid });
    },
    mergeIndividuals(event) {
      const droppedIndividuals = JSON.parse(
        event.dataTransfer.getData("individuals")
      );
      const uuids = droppedIndividuals.map(individual => individual.uuid);
      this.$emit("merge", [this.uuid, ...uuids]);
    },
    enrollIndividual(event) {
      const organization = event.dataTransfer.getData("organization");
      this.$emit("enroll", organization);
    }
  }
};
</script>
<style lang="scss" scoped>
@import "../styles/index.scss";

.v-list-item--three-line .v-list-item__avatar {
  font-size: 0.8rem;
}
</style>
