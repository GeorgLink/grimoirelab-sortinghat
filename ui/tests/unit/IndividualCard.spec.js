import { shallowMount, mount } from "@vue/test-utils";
import Vue from "vue";
import Vuetify from "vuetify";
import IndividualCard from "@/components/IndividualCard";
import * as Mutations from "@/apollo/mutations";

Vue.use(Vuetify);

const lockResponseMocked = {
  data: {
    lock: {
      uuid: "375458370ac0323bfb2e5a153e086551ef628d53",
      individual: {
        isLocked: true,
        __typename: "IndividualType"
      },
      __typename: "Lock"
    }
  }
};

const unlockResponseMocked = {
  data: {
    unlock: {
      uuid: "375458370ac0323bfb2e5a153e086551ef628d53",
      individual: {
        isLocked: false,
        __typename: "IndividualType"
      },
      __typename: "Unlock"
    }
  }
};

const errorResponseMocked = {
  errors: [
    {
      message: "abc not found in the registry",
      locations:[{line:2,column:3}],
      path:["lock"],
      extensions:{
        code:9
      }
    }],
    data:{
      lock:null
    }};

describe("IndividualCard", () => {
  test("Mock query for lockIndividual", async () => {
    const mutate = jest.fn(() => Promise.resolve(lockResponseMocked));
    const wrapper = shallowMount(IndividualCard, {
      Vue,
      mocks: {
        $apollo: {
          mutate
        }
      },
      propsData: {
        name: 'Test',
        uuid: "375458370ac0323bfb2e5a153e086551ef628d53",
        isLocked: false
      }
    });

    const response = await Mutations.lockIndividual(wrapper.vm.$apollo, wrapper.vm.uuid);

    expect(mutate).toBeCalled();
    expect(wrapper.element).toMatchSnapshot();
    });

    test("Mock query for unlockIndividual", async () => {
      const mutate = jest.fn(() => Promise.resolve(unlockResponseMocked));
      const wrapper = shallowMount(IndividualCard, {
        Vue,
        mocks: {
          $apollo: {
            mutate
          }
        },
        propsData: {
          name: 'Test',
          uuid: "375458370ac0323bfb2e5a153e086551ef628d53",
          isLocked: true
        }
      });

      const response = await Mutations.unlockIndividual(wrapper.vm.$apollo, wrapper.vm.uuid);

      expect(mutate).toBeCalled();
      expect(wrapper.element).toMatchSnapshot();
      });

      test("Lock individual on click", async () => {
        const mutate = jest.fn(() => Promise.resolve(lockResponseMocked));
        const lockIndividual = jest.spyOn(Mutations, 'lockIndividual');
        const wrapper = mount(IndividualCard, {
          Vue,
          mocks: {
            $apollo: {
              mutate
            }
          },
          propsData: {
            name: 'Test',
            uuid: "375458370ac0323bfb2e5a153e086551ef628d53",
            isLocked: false
          }
        });

        wrapper.find('button').trigger('click');

        expect(mutate).toBeCalled();
        expect(lockIndividual).toHaveBeenCalledWith(wrapper.vm.$apollo, "375458370ac0323bfb2e5a153e086551ef628d53");
        await Vue.nextTick();
        expect(wrapper.vm.locked).toBe(true);
        await Vue.nextTick();
        expect(wrapper.find('.mdi-lock').exists()).toBe(true);
        expect(wrapper.element).toMatchSnapshot();
      });

      test("Unlock individual on click", async () => {
        const mutate = jest.fn(() => Promise.resolve(unlockResponseMocked));
        const unlockIndividual = jest.spyOn(Mutations, 'unlockIndividual');
        const wrapper = mount(IndividualCard, {
          Vue,
          mocks: {
            $apollo: {
              mutate
            }
          },
          propsData: {
            name: 'Test',
            uuid: "375458370ac0323bfb2e5a153e086551ef628d53",
            isLocked: true
          }
        });

        wrapper.find('button').trigger('click');

        expect(mutate).toBeCalled();
        expect(unlockIndividual).toHaveBeenCalledWith(wrapper.vm.$apollo, "375458370ac0323bfb2e5a153e086551ef628d53");
        await Vue.nextTick();
        expect(wrapper.vm.locked).toBe(false);
        await Vue.nextTick();
        expect(wrapper.find('.mdi-lock-open-outline').exists()).toBe(true);
        expect(wrapper.element).toMatchSnapshot();
      });

      test("Show snackbar on error", async () => {
        const mutate = jest.fn(() => Promise.resolve(errorResponseMocked));
        const wrapper = mount(IndividualCard, {
          Vue,
          mocks: {
            $apollo: {
              mutate
            }
          },
          propsData: {
            name: 'Test',
            uuid: "abc",
            isLocked: true
          }
        });

        wrapper.find('button').trigger('click');

        expect(mutate).toBeCalled();
        await Vue.nextTick();
        expect(wrapper.vm.snackbar.text).toBe("abc not found in the registry");
        await Vue.nextTick();
        expect(wrapper.find('.v-snack').exists()).toBe(true);
        expect(wrapper.element).toMatchSnapshot();
      });
  });
