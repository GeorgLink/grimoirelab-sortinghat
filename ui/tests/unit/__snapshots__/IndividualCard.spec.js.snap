// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IndividualCard Lock individual on click 1`] = `
<div
  class="mx-auto v-card v-card--link v-card--raised v-sheet theme--light"
  tabindex="0"
>
  <div
    class="grow v-list-item theme--light"
    tabindex="-1"
  >
    <div
      class="v-avatar v-list-item__avatar grey"
      style="height: 40px; min-width: 40px; width: 40px;"
    >
      
      T
    
    </div>
     
    <div
      class="v-list-item__content"
    >
      <div
        class="v-list-item__title"
      >
        Test
      </div>
       
      <div
        class="v-list-item__subtitle"
      />
    </div>
     
    <div
      class="v-list-item__action"
    >
      <span
        class="v-list-item__action-text"
      >
        <!---->
         
        <button
          class="v-btn v-btn--flat v-btn--icon v-btn--round v-btn--text theme--light v-size--default"
          type="button"
        >
          <span
            class="v-btn__content"
          >
            <i
              aria-hidden="true"
              class="v-icon notranslate mdi mdi-lock theme--light"
              style="font-size: 16px;"
            />
          </span>
        </button>
      </span>
    </div>
  </div>
    
  <!---->
</div>
`;

exports[`IndividualCard Mock query for lockIndividual 1`] = `
<v-card-stub
  class="mx-auto"
  loaderheight="4"
  raised="true"
  tag="div"
>
  <v-list-item-stub
    activeclass=""
    class="grow"
    tag="div"
  >
    <v-list-item-avatar-stub
      color="grey"
      size="40"
    >
      
      T
    
    </v-list-item-avatar-stub>
     
    <v-list-item-content-stub>
      <v-list-item-title-stub>
        Test
      </v-list-item-title-stub>
       
      <v-list-item-subtitle-stub />
    </v-list-item-content-stub>
     
    <v-list-item-action-stub>
      <v-list-item-action-text-stub>
        <!---->
         
        <v-btn-stub
          activeclass=""
          icon="true"
          tag="button"
          text="true"
          type="button"
        >
          <v-icon-stub
            small=""
          >
            
            mdi-lock-open-outline
          
          </v-icon-stub>
        </v-btn-stub>
      </v-list-item-action-text-stub>
    </v-list-item-action-stub>
  </v-list-item-stub>
    
  <v-snackbar-stub
    color="error"
    timeout="6000"
  >
    
    Error
  
  </v-snackbar-stub>
</v-card-stub>
`;

exports[`IndividualCard Mock query for unlockIndividual 1`] = `
<v-card-stub
  class="mx-auto"
  loaderheight="4"
  raised="true"
  tag="div"
>
  <v-list-item-stub
    activeclass=""
    class="grow"
    tag="div"
  >
    <v-list-item-avatar-stub
      color="grey"
      size="40"
    >
      
      T
    
    </v-list-item-avatar-stub>
     
    <v-list-item-content-stub>
      <v-list-item-title-stub>
        Test
      </v-list-item-title-stub>
       
      <v-list-item-subtitle-stub />
    </v-list-item-content-stub>
     
    <v-list-item-action-stub>
      <v-list-item-action-text-stub>
        <!---->
         
        <v-btn-stub
          activeclass=""
          icon="true"
          tag="button"
          text="true"
          type="button"
        >
          <v-icon-stub
            small=""
          >
            
            mdi-lock
          
          </v-icon-stub>
        </v-btn-stub>
      </v-list-item-action-text-stub>
    </v-list-item-action-stub>
  </v-list-item-stub>
    
  <v-snackbar-stub
    color="error"
    timeout="6000"
  >
    
    Error
  
  </v-snackbar-stub>
</v-card-stub>
`;

exports[`IndividualCard Show snackbar on error 1`] = `
<div
  class="mx-auto v-card v-card--link v-card--raised v-sheet theme--light"
  tabindex="0"
>
  <div
    class="grow v-list-item theme--light"
    tabindex="-1"
  >
    <div
      class="v-avatar v-list-item__avatar grey"
      style="height: 40px; min-width: 40px; width: 40px;"
    >
      
      T
    
    </div>
     
    <div
      class="v-list-item__content"
    >
      <div
        class="v-list-item__title"
      >
        Test
      </div>
       
      <div
        class="v-list-item__subtitle"
      />
    </div>
     
    <div
      class="v-list-item__action"
    >
      <span
        class="v-list-item__action-text"
      >
        <!---->
         
        <button
          class="v-btn v-btn--flat v-btn--icon v-btn--round v-btn--text theme--light v-size--default"
          type="button"
        >
          <span
            class="v-btn__content"
          >
            <i
              aria-hidden="true"
              class="v-icon notranslate mdi mdi-lock theme--light"
              style="font-size: 16px;"
            />
          </span>
        </button>
      </span>
    </div>
  </div>
    
  <div
    class="v-snack v-snack--active v-snack--bottom v-snack-transition-enter v-snack-transition-enter-active"
  >
    <div
      class="v-snack__wrapper error"
      role="alert"
    >
      <div
        class="v-snack__content"
      >
        
    abc not found in the registry
  
      </div>
    </div>
  </div>
</div>
`;

exports[`IndividualCard Unlock individual on click 1`] = `
<div
  class="mx-auto v-card v-card--link v-card--raised v-sheet theme--light"
  tabindex="0"
>
  <div
    class="grow v-list-item theme--light"
    tabindex="-1"
  >
    <div
      class="v-avatar v-list-item__avatar grey"
      style="height: 40px; min-width: 40px; width: 40px;"
    >
      
      T
    
    </div>
     
    <div
      class="v-list-item__content"
    >
      <div
        class="v-list-item__title"
      >
        Test
      </div>
       
      <div
        class="v-list-item__subtitle"
      />
    </div>
     
    <div
      class="v-list-item__action"
    >
      <span
        class="v-list-item__action-text"
      >
        <!---->
         
        <button
          class="v-btn v-btn--flat v-btn--icon v-btn--round v-btn--text theme--light v-size--default"
          type="button"
        >
          <span
            class="v-btn__content"
          >
            <i
              aria-hidden="true"
              class="v-icon notranslate mdi mdi-lock-open-outline theme--light"
              style="font-size: 16px;"
            />
          </span>
        </button>
      </span>
    </div>
  </div>
    
  <!---->
</div>
`;
